pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,572 @@
+- A hexadecimal literal for base-16 numbers. These start with the prefix `0x`, followed by one or more digits between `0` and `9` or letters between `a` and `f`. Hexadecimal literals are not case-sensitive.
+- A binary literal for base-2 numbers. These start with the prefix `0b`, followed by one or more digits that are either `0` or `1`.
+
+Additionally, underscores can appear between any two characters of an integer literal.
+These underscores carry no semantic meaning and are discarded by the parser.
+
+Examples of valid integer literals:
+
+```slice
+123
+0xFF
+0b1010
+
+335_445_996
+0x_ab_cd_ef  // Same as 0xabcded
+0b0_______1  // Same as 0b01
+
+0x_0b1101  // hexadecimal literal of value 725249
+
+// All have a value of 0
+0
+0_00_0
+0b000
+0x0
+```
+
+{% callout type="note" %}
+Note that `-` signs are counted as separate tokens, and not as part of an integer literal.
+{% /callout %}
+
+The Slice grammar has 2 separate rules for unsigned and (possibly) signed integers.
+Slice does not support using a `+` sign to indicate an integer is positive, since all integers are positive by default.
+
+```ebnf {% showTitle=false %}
+Integer
+    : integer_literal
+    ;
+
+SignedInteger
+    : Integer
+    | "-" Integer
+    ;
+```
+
+### String literals
+
+String literals consist of any number of UTF8 characters, wrapped in a pair of double quotes.
+Any characters within a string literal are treated as opaque text by the parser.
+
+Slice supports escaping special characters within string literals, this allows you to include unprintable characters or literal double quote characters in your string. Characters can be escaped by prefixing them with a backslash.
+To write a literal backslash, you must escape it with another backslash: `\\`.
+
+When parsing, Slice will resolve any escaped characters, removing the `\` escape character where present.
+Apart from this, Slice performs no additional processing of string literals.
+
+Examples of valid string literals:
+
+```slice
+""
+"hello world"
+"the variable \"foo\" doesn't exist"
+"a backslash: '\\'"
+```
+
+### Attributes
+
+Slice supports two forms of attributes:
+- Local attributes: these consist of an attribute wrapped in a pair of single brackets.
+- File attributes: these consist of an attribute wrapped in a pair of double brackets.
+
+Local attributes can be applied to most Slice definitions by placing them directly before the definition.
+File attributes apply to an entire [slice file][slice-file] and must appear before any Slice definitions or module declarations.
+
+```ebnf {% showTitle=false %}
+FileAttribute
+    : "[[" Attribute "]]"
+    ;
+
+LocalAttribute
+    : "[" Attribute "]"
+    ;
+```
+
+An attribute consists of a directive, optionally followed by attribute arguments.
+A directive is syntactically equivalent to a [scoped identifier][identifier], but instead of Slice scopes, its scopes determine when the attribute is meaningful.
```

Comment:
I find the comparison with Slice scopes misleading since we use different naming conventions: module names are by convention in Pascal case while attribute scopes are lower case.

Reply:
The conventions are different, but I don't think conventions are technically relevant in the reference.
Even if we suggest different conventions, they are still syntactically equivalent.

This sentence isn't about advertising "use these like identifiers",
it's for saying "we're not going to restate what this is, since it's the same as this other thing; click here"